src/playground-code-editor.ts

```diff
@@ -384,17 +411,252 @@ export class PlaygroundCodeEditor extends LitElement {
+    changeObject: EditorChange
+  ) {
+    if (!this._currentFiletypeSupportsCompletion()) return;
+
+    const previousToken = editorInstance.getTokenAt(changeObject.from);
+    const tokenUnderCursor = this.tokenUnderCursor.string.trim();
+    const tokenUnderCursorAsString = tokenUnderCursor.trim();
+    // To help reduce round trips to a language service or a completion provider, we
+    // are providing a flag if the completion is building on top of the earlier recommendations.
+    // If the flag is true, the completion system can just filter the already stored
+    // collection of completions again with the more precise input.
+    // On deletion events, we want to query the LS again, since we might be in a new context after
+    // removing characters from our code.
+    const isInputEvent = changeObject.origin === '+input';
+    const isRefinement =
+      (tokenUnderCursor.length > 1 || previousToken.string === '.') &&
+      isInputEvent;
+    const changeWasCodeCompletion = changeObject.origin === 'complete';
+
+    if (tokenUnderCursorAsString.length <= 0) return;
+
+    if (changeWasCodeCompletion) {
+      // If the case that the user triggered a code completion,
+      // we want to empty out the completions until
+      // a letter is input.
+      this.completions = [];
+      return;
+    }
+
+    const id = ++this._currentCompletionRequestId;
+    this.dispatchEvent(
+      new CustomEvent('request-completions', {
+        detail: {
+          isRefinement,
+          fileName: this.fileName,
+          fileContent: this.value,
+          tokenUnderCursor,
+          cursorIndex: this.cursorIndex,
+          provideCompletions: (completions: EditorCompletion[]) =>
+            this._onCompletionsProvided(id, completions),
+        },
+      })
+    );
+  }
+
+  private _onCompletionsProvided(id: number, completions: EditorCompletion[]) {
+    // To prevent race conditioning, check that the completions provided
+    // are from the latest completions request.
+    if (id !== this._currentCompletionRequestId) {
+      return;
+    }
+
+    this.completions = completions;
+  }
+
+  private _currentFiletypeSupportsCompletion() {
+    // Currently we are only supporting code completion for TS. Change
+    // this in a case that we start to support it for other languages too.
+    return this.type === 'ts';
+  }
+
   focus() {
     this._codemirrorEditable?.focus();
   }
 
+  private _hintFunction(
+    cm: Editor
+  ): Hints | null | undefined | PromiseLike<Hints | null | undefined> {
+    const cursorPosition = cm.getCursor('start');
+    const token = cm.getTokenAt(cursorPosition);
+    const lineNumber = cursorPosition.line;
+
+    const hintList = this._completionsAsHints();
+
+    const hints = {
+      from: {line: lineNumber, ch: token.start} as Position,
+      to: {line: lineNumber, ch: token.end} as Position,
+      list: hintList,
+    } as Hints;
+
+    CodeMirror.on(
+      hints,
+      'select',
```

Comment:
Are these event listeners going to be garbage collected correctly? Looks like we're creating a new one every time the completions menu pops up? Is this delegating to `addEventListener` under the hood, or is it a bespoke codemirror event system?

Related: if we were writing our own completions UI, it would probably be a single container element, with a single event listener, which just gets updated and shown/hidden. Is codemirror's implementation creating a new element from scratch every time?

Reply:
AFAIK from my research it attaches a eventlistener with `addEventListener` to the `ul`  element containing all of the completion items. Then as it closes, it removes the whole element from DOM.

So there is only one eventlistener per action, no matter the amount of completion items listed.

So...

 > Is this delegating to addEventListener under the hood, or is it a bespoke codemirror event system?
 
 It is attaching a single eventlistner per action (e.g. dblclick, click, etc.) to the UL element wrapping the list
 
 >  Is codemirror's implementation creating a new element from scratch every time?
 
  Yeah. It creates it from scratch, appends `li` items inside it and as it's closed, it rids of it with parentNode.removeChild(hints)